### DISCUSSION TITLE
The name of action trigger by persist middleware is always anonymous in redux devtools

### DISCUSSION THREAD

**Discussion by credred** [pmndrs/zustand#2274] [Q&A] [answered]
I am using devtools and persist middleware. I noticed that when the persist middleware was initialized, the set method only passed two parameters. Can I add configuration and pass more parameters?

https://github.com/pmndrs/zustand/blob/517524d2b05843c6a26d2713af08cc0ca1f62f3c/src/middleware/persist.ts#L476

**Comment by dai-shi** [comment]
Does it look like this?
https://github.com/pmndrs/zustand/blob/517524d2b05843c6a26d2713af08cc0ca1f62f3c/tests/middlewareTypes.test.tsx#L380-L392

**Reply by credred** [reply]
When hydrate is triggered for the first time or rehyrate is called manually, the action seen by redux devtools is anonymous.
I want to customize this action name, such as restoreFromStorage.
This is a mini reproduce link: https://stackblitz.com/edit/typescript-fd3pbm?file=index.html,index.ts,node_modules%2Fzustand%2Fmiddleware%2Fpersist.d.ts

**Reply by dai-shi** [reply]
I see. Do you have any idea for the fix?

**Reply by credred** [reply]
What I can think of is providing an options to customize the set, such as `options.customSet(stateFromStorage as S, true)`. But people like me have less need for set customization, which may lead to more and more middleware options.

**Reply by dai-shi** [reply]
I think I misunderstood something. Can you share the reproduction again? The link doesn't have a content.

**Reply by credred** [reply]
Sorry for affecting your time, the link has been updated.

**Reply by dai-shi** [reply]
got it. okay, this is tricky as there are two middlewares. It feels like TS typing is almost impossible.

are you aware that you can customize anonymous type?
https://github.com/pmndrs/zustand/blob/4be1e9eb4a435e5821d5ee24441a529881607f79/src/middleware/devtools.ts#L65

**Reply by credred** [reply]
I know this option. Suppose I set anonymousActionType to restoreFromStorage. In the following example, after the inc method is executed, the redux devtool may mistakenly think that this is executed by the persist middleware.

```

const store = createStore<CounterState>()(
  devtools(
    persist(
      (set, get) => ({
        count: 1,
        inc: () => set({ count: get().count + 1 }, false),
      }),
      {
        name: 'count',
        partialize: (state) => {
          count: state.count;
        },
      }
    ),
    { name: 'prefix' }
  )
);

setTimeout(() => {
  // action name is restoreFromStorage also
  store.getState().inc()
}, 1000)

```

**Reply by dai-shi** [reply]
Yeah, you need to avoid anonymous:
```diff
-         inc: () => set({ count: get().count + 1 }, false),
+         inc: () => set({ count: get().count + 1 }, false, 'inc'),
```

**Reply by credred** [reply]
Once someone in the team forgets to write the action name, others will be confused when looking at redux devtools...

**Reply by credred** [reply]
If we see anonymous, we will know that someone forgot to write the action name. If we see restoreFromStorage, we cannot distinguish whether it is forgetting to write the action name or whether the persist middleware is restoring the data.

**Reply by dai-shi** [reply]
it makes sense. it's just hard to implement it generally cleanly. for now, you can try having your own `persist` middleware.

**Comment by dai-shi** [comment, marked as answer]
One idea for the implementation, although it's bad developer experience:
```js
store.devtools.withAnonymousActionType('rehydrate', () => {
  store.persist.rehydrate();
});
```

**Reply by credred** [reply]
This is a good idea and this solves my problem. Thank you for your patient answer

**Reply by dai-shi** [reply]
It's just an idea. We don't to plan to implement it yet.
